Fix for #1621: only concat x1d spectra from MIRI

bmorris3 opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our code of conduct,
https://github.com/spacetelescope/jdaviz/blob/main/CODE_OF_CONDUCT.md . -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable viz component(s). -->

This pull request is to address #1621. By default on main, specviz will read multi-order x1d spectra into a `SpectrumList`, and then cobble together every order into a concatenated 1D spectrum. This behavior may have made sense for MIRI data products but doesn't give us a sensible default view for NIRISS. This PR adds ~a small check for the instrument: if the observations were recorded by MIRI jdaviz will use the old behavior~ a check if the various HDUs originated from the same FITS file. If the loaded orders are from the same FITS file, specviz will compute a "combined" spectrum with all spectral channels. 

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #1621 

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge.

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

kecnry commented on jdaviz/configs/specviz/plugins/parsers.py:
```diff
@@ -117,28 +117,34 @@ def specviz_spectrum1d_parser(app, data, data_label=None, format=None, show_in_v
+                app.add_data(spec, label)
+                if show_in_viewer:
+                    app.add_data_to_viewer("spectrum-viewer", label)
+            else:
+                # show only the first FITS extension if the observations
+                # were produced by instrument other than MIRI
+                if show_in_viewer:
+                    app.add_data_to_viewer("spectrum-viewer", data_label[0])
```
does this block need an `add.add_data call` or is that handled somewhere else?

bmorris3 replied:
Pretty sure that's done [up here](https://github.com/spacetelescope/jdaviz/blob/75eb05f3e48feb653df5c93f2fe4f71d4056bcc6/jdaviz/configs/specviz/plugins/parsers.py#L105).

bmorris3 replied:
Exactly! 👍

pllim replied:
Is this necessary? `add_data_to_viewer` is already done in L118 above.

bmorris3 replied:
My understanding was that you still needed to call `add_data` before you call `add_data_to_viewer` – is that right?

pllim replied:
Yes, I think so...

pllim commented on jdaviz/configs/specviz/plugins/parsers.py:
```diff
@@ -117,28 +117,34 @@ def specviz_spectrum1d_parser(app, data, data_label=None, format=None, show_in_v
-            app.add_data(spec, label)
-            if show_in_viewer:
-                app.add_data_to_viewer("spectrum-viewer", label)
+            if data[0].meta.get("INSTRUME") == "MIRI":
```
I can see two use cases where this logic might fail:

* User provides a SpectrumList with mixed instruments and the first one happens to be MIRI and user does not want the composite spectrum.
* User provides a SpectrumList for some telescope we never heard of but user wants a composite spectrum.

Is there a way to generalize this solution to be less JWST specific?

pllim replied:
What about a new keyword so user can explicitly ask for composite spectrum or not. And if it is not given and detector is MIRI, then do the thing you do.

bmorris3 replied:
@pllim Sorry it took me a while to come back to. I implemented your suggestion in [0fb4653](https://github.com/spacetelescope/jdaviz/pull/1659/commits/0fb46536b0df1dad08c9a1d26392f90791e991da).

pllim reviewed (commented):
Not bad for your first PR. 😸 👏 

Specviz2d also uses this parser. How will this change affect Specviz2d ?

Need a change log but that can wait if you are afraid of merge conflicts.

Does any user facing doc needs to be updated?

bmorris3 commented on jdaviz/configs/specviz/plugins/parsers.py:
```diff
@@ -118,33 +117,82 @@ def specviz_spectrum1d_parser(app, data, data_label=None, format=None, show_in_v
+            # group them by their original FITS filenames and add their combined
+            # 1D spectrum to the DataCollection.
+            unique_files = group_spectra_by_filename(app.data_collection)
+            print(unique_files)
```
Yikes, meant to delete this.

bmorris3 commented on jdaviz/configs/specviz/plugins/parsers.py:
```diff
@@ -118,33 +117,82 @@ def specviz_spectrum1d_parser(app, data, data_label=None, format=None, show_in_v
+            # If >1 spectra in the list were opened from the same FITS file,
+            # group them by their original FITS filenames and add their combined
+            # 1D spectrum to the DataCollection.
+            unique_files = group_spectra_by_filename(app.data_collection)
```
This logic adds a composite spectrum to the data collection for every unique FITS file loaded via, for example, `SpectrumList.read`. This approach allows users to pass in heterogeneous `SpectrumList`s originating from multiple instruments, and will concatenate the spectra with multiple orders only when they originate from the same data product.

pllim replied:
I don't understand how this plays together with this line above:

https://github.com/spacetelescope/jdaviz/blob/eb9cfd5e6fddd690e10460f8ac5f90692bb548d3/jdaviz/configs/specviz/plugins/parsers.py#L96

You sure we're not duplicating the initially loaded stuff here?

bmorris3 replied:
I'm not sure I follow. The `for i, spec in enumerate(data)` block is adding the individual spectra one at a time to the data collection and/or viewer. The `if concat_by_file and isinstance(data, SpectrumList)` block computes "combined"/"concatenated" spectra for the contents of each loaded FITS file, and adds the combined spectra to the collection/viewer. I don't think there's duplication here.

pllim replied:
Ah, gotcha. I got lost in the code, so I appreciate you made it clear to me.

bmorris3 commented:
The CI test that is [failing above](https://github.com/spacetelescope/jdaviz/actions/runs/3092593668/jobs/5004039309#step:5:393) does not fail anymore when using the dev version of specutils, since https://github.com/astropy/specutils/pull/982 was merged. There are a few ways to approach this failing test. Should I: 
* increment the required specutils version in `setup.cfg`
* add a `@pytest.mark.skipif` decorator to the test which runs different checks depending on the specutils version
* something else?

duytnguyendtn commented:
> increment the required specutils version in setup.cfg

I believe this is the right answer here, as opposed to having the tests skip. Interested to hear if anyone disagrees though!

kecnry commented:
agreed, I don't think there is a huge risk of major conflicts if this sits unmerged until specreduce is released and the new version pinned as part of this PR.

rosteen commented:
Yup, I was going to say, pin to the new specutils version - but I have to release that first 😆

kecnry commented on jdaviz/configs/specviz/plugins/parsers.py:
```diff
@@ -117,28 +116,83 @@ def specviz_spectrum1d_parser(app, data, data_label=None, format=None, show_in_v
+            # group them by their original FITS filenames and add their combined
+            # 1D spectrum to the DataCollection.
+            unique_files = group_spectra_by_filename(app.data_collection)
+            print(unique_files)
```
Let's not forget then!

```suggestion
```

kecnry commented on jdaviz/configs/specviz/tests/test_helper.py:
```diff
@@ -248,7 +254,7 @@ def test_load_spectrum_list_directory(tmpdir, specviz_helper):
 
     with pytest.warns(UserWarning, match='SRCTYPE is missing or UNKNOWN in JWST x1d loader'):
         specviz_helper.load_spectrum(data_path)
-    assert len(specviz_helper.app.data_collection) == 3
+    assert len(specviz_helper.app.data_collection) == 41
```
is this actually expected to be 41?  This same test has also changed on main to give a length of 40 (failing when expecting 3)... 🤔

pllim replied:
Did someone dump a bunch of new files into the ZIP file we download for the test? 😱

bmorris3 replied:
Thanks for the summary @kecnry. The new behavior after this PR is implemented is to load all of the available extensions within the directory (39 total), and make concatenated spectra from each FITS file (2 additional), making 41. Though I'm not 100% sure we want that behavior in this instance, since these are NIRISS observations that we may not want to have concatenated.

pllim replied:
Contents of the ZIP file (minus that random notebook that we don't need):

```python
['.\\canucs_basic_F150W_direct_dit1_i2d.fits',
 '.\\canucs_basic_F150W_WFSSC_dit1_cal.fits',
 '.\\canucs_basic_F150W_WFSSC_dit1_x1d.fits',
 '.\\canucs_basic_F150W_WFSSR_dit1_cal.fits',
 '.\\canucs_basic_F150W_WFSSR_dit1_x1d.fits']
```

For x1d , here is the FITS info:

```
Filename: .\canucs_basic_F150W_WFSSC_dit1_x1d.fits
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  PRIMARY       1 PrimaryHDU     259   ()
  1  EXTRACT1D     1 BinTableHDU     47   107R x 9C   [D, D, D, D, D, J, D, D, D]
  2  EXTRACT1D     2 BinTableHDU     47   42R x 9C   [D, D, D, D, D, J, D, D, D]
  3  EXTRACT1D     3 BinTableHDU     47   110R x 9C   [D, D, D, D, D, J, D, D, D]
  4  EXTRACT1D     4 BinTableHDU     47   103R x 9C   [D, D, D, D, D, J, D, D, D]
  5  EXTRACT1D     5 BinTableHDU     47   107R x 9C   [D, D, D, D, D, J, D, D, D]
  6  EXTRACT1D     6 BinTableHDU     47   108R x 9C   [D, D, D, D, D, J, D, D, D]
  7  EXTRACT1D     7 BinTableHDU     47   103R x 9C   [D, D, D, D, D, J, D, D, D]
  8  EXTRACT1D     8 BinTableHDU     47   107R x 9C   [D, D, D, D, D, J, D, D, D]
  9  EXTRACT1D     9 BinTableHDU     47   93R x 9C   [D, D, D, D, D, J, D, D, D]
 10  EXTRACT1D    10 BinTableHDU     47   105R x 9C   [D, D, D, D, D, J, D, D, D]
 11  EXTRACT1D    11 BinTableHDU     47   108R x 9C   [D, D, D, D, D, J, D, D, D]
 12  EXTRACT1D    12 BinTableHDU     47   111R x 9C   [D, D, D, D, D, J, D, D, D]
 13  EXTRACT1D    13 BinTableHDU     47   110R x 9C   [D, D, D, D, D, J, D, D, D]
 14  EXTRACT1D    14 BinTableHDU     47   104R x 9C   [D, D, D, D, D, J, D, D, D]
 15  EXTRACT1D    15 BinTableHDU     47   111R x 9C   [D, D, D, D, D, J, D, D, D]
 16  EXTRACT1D    16 BinTableHDU     47   106R x 9C   [D, D, D, D, D, J, D, D, D]
 17  EXTRACT1D    17 BinTableHDU     47   101R x 9C   [D, D, D, D, D, J, D, D, D]
 18  EXTRACT1D    18 BinTableHDU     47   58R x 9C   [D, D, D, D, D, J, D, D, D]
 19  EXTRACT1D    19 BinTableHDU     47   111R x 9C   [D, D, D, D, D, J, D, D, D]
 20  EXTRACT1D    20 BinTableHDU     47   108R x 9C   [D, D, D, D, D, J, D, D, D]
 21  ASDF          1 BinTableHDU     11   1R x 1C   [107650B]
```

```
Filename: .\canucs_basic_F150W_WFSSR_dit1_x1d.fits
No.    Name      Ver    Type      Cards   Dimensions   Format
  0  PRIMARY       1 PrimaryHDU     259   ()
  1  EXTRACT1D     1 BinTableHDU     47   3R x 9C   [D, D, D, D, D, J, D, D, D]
  2  EXTRACT1D     2 BinTableHDU     47   19R x 9C   [D, D, D, D, D, J, D, D, D]
  3  EXTRACT1D     3 BinTableHDU     47   78R x 9C   [D, D, D, D, D, J, D, D, D]
  4  EXTRACT1D     4 BinTableHDU     47   107R x 9C   [D, D, D, D, D, J, D, D, D]
  5  EXTRACT1D     5 BinTableHDU     47   110R x 9C   [D, D, D, D, D, J, D, D, D]
  6  EXTRACT1D     6 BinTableHDU     47   107R x 9C   [D, D, D, D, D, J, D, D, D]
  7  EXTRACT1D     7 BinTableHDU     47   105R x 9C   [D, D, D, D, D, J, D, D, D]
  8  EXTRACT1D     8 BinTableHDU     47   112R x 9C   [D, D, D, D, D, J, D, D, D]
  9  EXTRACT1D     9 BinTableHDU     47   107R x 9C   [D, D, D, D, D, J, D, D, D]
 10  EXTRACT1D    10 BinTableHDU     47   109R x 9C   [D, D, D, D, D, J, D, D, D]
 11  EXTRACT1D    11 BinTableHDU     47   109R x 9C   [D, D, D, D, D, J, D, D, D]
 12  EXTRACT1D    12 BinTableHDU     47   116R x 9C   [D, D, D, D, D, J, D, D, D]
 13  EXTRACT1D    13 BinTableHDU     47   103R x 9C   [D, D, D, D, D, J, D, D, D]
 14  EXTRACT1D    14 BinTableHDU     47   111R x 9C   [D, D, D, D, D, J, D, D, D]
 15  EXTRACT1D    15 BinTableHDU     47   104R x 9C   [D, D, D, D, D, J, D, D, D]
 16  EXTRACT1D    16 BinTableHDU     47   101R x 9C   [D, D, D, D, D, J, D, D, D]
 17  EXTRACT1D    17 BinTableHDU     47   109R x 9C   [D, D, D, D, D, J, D, D, D]
 18  EXTRACT1D    18 BinTableHDU     47   107R x 9C   [D, D, D, D, D, J, D, D, D]
 19  EXTRACT1D    19 BinTableHDU     47   108R x 9C   [D, D, D, D, D, J, D, D, D]
 20  ASDF          1 BinTableHDU     11   1R x 1C   [101454B]
```

So 20 + 19 = 39. And I guess the other 1 or 2 links go to `i2d` and `cal` linking? When you load this dataset into Jdaviz manually, does it behave correctly? I assume each extension is a different slit, so when you go through the table, do things change accordingly?

kecnry replied:
ok, then since this is an expected change by the change to specutils, #1752 updates the test to 40 (so that other PRs can rebase on that in the meantime until this change is merged) and you'll then just have to keep this change to 41 after rebasing on top of that.

pllim replied:
Oh wait, I thought this was Mosviz. But now I see it is Specviz. How did 39 `EXTRACT1D` spectra create 40-41 links? Shouldn't it be 38? 🤯

bmorris3 replied:
Here are the data that it loads/concatenates: 
```
DataCollection (41 data sets)
	  0: NIRISS_for_parser_p0171 0
	  1: NIRISS_for_parser_p0171 1
	  2: NIRISS_for_parser_p0171 2
	  3: NIRISS_for_parser_p0171 3
	  4: NIRISS_for_parser_p0171 4
	  5: NIRISS_for_parser_p0171 5
	  6: NIRISS_for_parser_p0171 6
	  7: NIRISS_for_parser_p0171 7
	  8: NIRISS_for_parser_p0171 8
	  9: NIRISS_for_parser_p0171 9
	 10: NIRISS_for_parser_p0171 10
	 11: NIRISS_for_parser_p0171 11
	 12: NIRISS_for_parser_p0171 12
	 13: NIRISS_for_parser_p0171 13
	 14: NIRISS_for_parser_p0171 14
	 15: NIRISS_for_parser_p0171 15
	 16: NIRISS_for_parser_p0171 16
	 17: NIRISS_for_parser_p0171 17
	 18: NIRISS_for_parser_p0171 18
	 19: NIRISS_for_parser_p0171 19
	 20: NIRISS_for_parser_p0171 20
	 21: NIRISS_for_parser_p0171 21
	 22: NIRISS_for_parser_p0171 22
	 23: NIRISS_for_parser_p0171 23
	 24: NIRISS_for_parser_p0171 24
	 25: NIRISS_for_parser_p0171 25
	 26: NIRISS_for_parser_p0171 26
	 27: NIRISS_for_parser_p0171 27
	 28: NIRISS_for_parser_p0171 28
	 29: NIRISS_for_parser_p0171 29
	 30: NIRISS_for_parser_p0171 30
	 31: NIRISS_for_parser_p0171 31
	 32: NIRISS_for_parser_p0171 32
	 33: NIRISS_for_parser_p0171 33
	 34: NIRISS_for_parser_p0171 34
	 35: NIRISS_for_parser_p0171 35
	 36: NIRISS_for_parser_p0171 36
	 37: NIRISS_for_parser_p0171 37
	 38: NIRISS_for_parser_p0171 38
	 39: Combined [canucs_basic_F150W_WFSSR_dit1_x1d.fits]
	 40: Combined [canucs_basic_F150W_WFSSC_dit1_x1d.fits]
```

camipacifici commented:
Wait, I understand this is just for a test, but these are extracted spectra of different sources. They should never be concatenated.

pllim commented on jdaviz/configs/specviz/tests/test_helper.py:
```diff
@@ -248,7 +254,7 @@ def test_load_spectrum_list_directory(tmpdir, specviz_helper):
     with pytest.warns(UserWarning, match='SRCTYPE is missing or UNKNOWN in JWST x1d loader'):
         specviz_helper.load_spectrum(data_path)
-    assert len(specviz_helper.app.data_collection) == 3
+    assert len(specviz_helper.app.data_collection) == 41
```
According to https://github.com/spacetelescope/jdaviz/pull/1659#issuecomment-1284504407 , no combined anything should be produced here. 🤷 

```suggestion
    assert len(specviz_helper.app.data_collection) == 39
```

bmorris3 replied:
In the latest commits I've changed this behavior (again). See below!

bmorris3 commented on jdaviz/configs/specviz/helper.py:
```diff
@@ -39,7 +39,7 @@ def __init__(self, *args, **kwargs):
 
     def load_spectrum(self, data, data_label=None, format=None, show_in_viewer=True,
-                      spectrum_viewer_reference_name=None):
+                      spectrum_viewer_reference_name=None, concat_by_file=False):
```
I've now added this kwarg which allows users like @orifox to get their desired behavior, but only when they ask for it. By default, no concatenated spectrum is assembled or loaded.

kecnry replied:
`load_spectrum` doesn't have any API docstring... but maybe it should (I did see that you added the new entry in the parser docstring, but not sure how easy that will be for users to find)?

Should this change (even though its an _addition_) warrant pushing back to 3.1 instead of 3.0.3?

bmorris3 replied:
> load_spectrum doesn't have any API docstring

I was a bit surprised when I edited the call signature that there was no docstring to append it to. Do you think I should write one up in this PR?

> 3.1 instead of 3.0.3

The only reason I labeled earlier was because @orifox has wanted this for a long time 😅

pllim replied:
Ori knows how to install dev version. 😉 

I don't think we should backport either.

pllim replied:
I also don't see any new docstring here, as requested above. I agree that would be nice to have even if just to set `load_spectrum.__doc__ = load_data.__doc__` (see if that is even possible and if they are indeed completely equivalent, if not just type it out).

bmorris3 commented on jdaviz/configs/specviz/tests/test_helper.py:
```diff
@@ -262,6 +276,24 @@ def test_load_spectrum_list_directory(tmpdir, specviz_helper):
+    # This feature is intended for MIRI spectra, but is tested on NIRISS spectra here:
+    with pytest.warns(UserWarning, match='SRCTYPE is missing or UNKNOWN in JWST x1d loader'):
+        specviz_helper.load_spectrum(data_path, concat_by_file=True)
+    assert len(specviz_helper.app.data_collection) == 41
```
This new test concatenates the spectra within each file into one combined spectrum per file.

pllim replied:
> # This feature is intended for MIRI spectra, but is tested on NIRISS spectra here

That is very confusing. Can't we have a less confusing test data?

bmorris3 replied:
I understand and agree that it's confusing. I used the NIRISS spectra here only because we have them uploaded to Box and ready to go. Would it be better to find and upload some MIRI observations instead?

pllim replied:
Okay, I think I understand now. The key is really `concat_by_file=True` no matter what instrument. Whether it is intended for MIRI or not, there is actually no code to enforce, so the comment is actually confusing as it tries to explain the original intent but not really what is happening... I guess I can let this go if you want to leave it as-is.

bmorris3 commented on jdaviz/configs/specviz/tests/test_helper.py:
```diff
@@ -38,15 +40,23 @@ def test_load_spectrum1d(self):
+        # now load three more spectra from a SpectrumList, without labels
         self.spec_app.load_spectrum(self.spec_list)
-        assert len(self.spec_app.app.data_collection) == 5
+        assert len(self.spec_app.app.data_collection) == 4
```
I don't understand why this test was working before – `self.spec_list` is a list of three spectra. There is one spectrum [already loaded](https://github.com/spacetelescope/jdaviz/pull/1659/files#diff-5ef5cbec11fb8f2fb8f31547e9995778108c86e1f383e9676e2b9ad52830c1bfR28), so there should be 4 in the collection here.

pllim replied:
This reuses `self.spec_app` modified in other tests. Maybe the number depends on which test ran first? Is it not cleaner to have each test method use their own `specviz_helper`?

bmorris3 replied:
Well now that I understand the problem the answer here is pretty clear 😅. The point of making this PR was to prevent Specviz from concatenating spectra and creating a new data item by default. That "extra" data item that made this test pass before was that concatenated spectrum. This PR changes the default behavior of loading `SpectrumList`s, so the concatenated spectrum is no longer computed or loaded by default. So the length of the data collection indeed should change from five (one from the `setup_class` method, three from `load_spectrum(spec_list)`, and one from the concatenation of the previous three spectra) to four (skipping the concatenation).

pllim replied:
Okay, in that case, maybe no need to refactor the tests... yet.

bmorris3 commented:
@camipacifici: I totally agree those files shouldn't be concatenated. I've added a keyword argument to `load_spectrum` in the latest few commits which allows the user to turn the concat feature on (default is off). I also added a new test to check that the toggle works as intended on these test data for NIRISS, even though in real life we wouldn't concatenate those data.

bmorris3 commented on jdaviz/configs/specviz/tests/test_helper.py:
```diff
@@ -248,12 +258,16 @@ def test_load_spectrum_list_directory(tmpdir, specviz_helper):
     # NOTE: the length was 3 before specutils 1.9 (https://github.com/astropy/specutils/pull/982)
     SPECUTILS_LT_1_9 = Version(specutils.__version__) < Version('1.9.0')
-    expected_len = 40 if not SPECUTILS_LT_1_9 else 3
+    expected_len = 39 if not SPECUTILS_LT_1_9 else 3
```
This needs to be revised because we don't expect a concatenated spectrum to be loaded by default, so we expected the length [described here](https://github.com/spacetelescope/jdaviz/pull/1659/files#diff-5ef5cbec11fb8f2fb8f31547e9995778108c86e1f383e9676e2b9ad52830c1bfR291).

kecnry replied:
If you're also bumping specutils here, I think you can safely get rid of the if statement entirely as part of this PR.

bmorris3 replied:
Ok great, I wasn't sure if this had to do with backporting?

pllim replied:
If we are bumping specutils minversion, should we even backport?

kecnry commented on jdaviz/configs/specviz/plugins/parsers.py:
```diff
@@ -27,6 +28,10 @@ def specviz_spectrum1d_parser(app, data, data_label=None, format=None, show_in_v
         `~specutils.Spectrum1D.read` io method.
     spectrum_viewer_reference_name : str
         Reference name for the viewer
+    concat_by_file : bool
+        If True, concatenate the extensions within each spectrum file
+        passed to the parser and add a concatenated spectrum to the
+        data collection.
```
is there any error/warning if the file does not contain multiple extensions?  It seems not from the code (which I think makes sense), but maybe the docstring should state that it is ignored otherwise?

bmorris3 replied:
I've edited this docstring in the latest commit.

kecnry commented on CHANGES.rst:
```diff
@@ -57,6 +57,9 @@ Mosviz
 Specviz
 ^^^^^^^
 
+- Only combine multi-order x1d spectra into a concatenated spectrum for MIRI (and not
+  for NIRISS, for example) [#1659]
```
this sounds like its automated, but this is currently entirely controlled by the user via `concat_by_file`, right?

bmorris3 replied:
I agree, editing the log now.

pllim replied:
Did the edit happen? I don't see any new commit.

pllim commented on jdaviz/configs/specviz/plugins/parsers.py:
```diff
@@ -117,28 +121,82 @@ def specviz_spectrum1d_parser(app, data, data_label=None, format=None, show_in_v
+    Parameters
+    ----------
+    dataset : `~glue.core.data_collection.DataCollection`
+        Collection of datasets
```
```suggestion
        Collection of datasets.
```

pllim commented on jdaviz/configs/specviz/plugins/parsers.py:
```diff
@@ -117,28 +121,82 @@ def specviz_spectrum1d_parser(app, data, data_label=None, format=None, show_in_v
+        filename = data.meta.get("FILENAME")
+
+        if filename not in spectra_to_combine:
+            spectra_to_combine[filename] = []
```
Would `defaultdict(list)` work better or is that not suited for your use case?

bmorris3 replied:
That's handy, I haven't used that before!

pllim commented on jdaviz/conftest.py:
```diff
@@ -156,6 +158,30 @@ def spectrum_collection(spectrum1d):
+    sc = []
+
+    for i in range(spectral_orders):
+        np.random.seed(42)
```
Do you mean to set the seed inside the loop? Won't all the spectra actually have the same values like this?

bmorris3 replied:
Yikes, thanks!

bmorris3 replied:
D'oh! Fixed.

pllim commented on jdaviz/conftest.py:
```diff
@@ -156,6 +158,30 @@ def spectrum_collection(spectrum1d):
+        sc.append(spectrum1d)
+
+    with warnings.catch_warnings():
+        warnings.simplefilter('ignore')
```
Why are we ignoring warnings here?

bmorris3 replied:
I built the test on tutorials elsewhere in the docs where this was necessary. It's not anymore, so I've removed it. 👍🏻

pllim commented on CHANGES.rst:
```diff
@@ -22,6 +22,8 @@ Specviz
 
 - Spectrum viewer now shows X and Y values under cursor. [#1759]
 
+- Switch to opt-in concatenation for multi-order x1d spectra [#1659]
```
```suggestion
- Switch to opt-in concatenation for multi-order x1d spectra. [#1659]
```

pllim commented on jdaviz/configs/specviz/helper.py:
```diff
@@ -39,27 +39,32 @@ def __init__(self, *args, **kwargs):
                                handler=self._redshift_listener)
 
     def load_spectrum(self, data, data_label=None, format=None, show_in_viewer=True,
-                      spectrum_viewer_reference_name=None):
```
Hmm you added `spectrum_viewer_reference_name` in #1681 but it looks unused. So I think it's ok to remove.

bmorris3 replied:
Correct, it should have been removed during revisions of #1681. Removing it now.

pllim commented on jdaviz/configs/specviz/plugins/parsers.py:
```diff
@@ -117,28 +120,78 @@ def specviz_spectrum1d_parser(app, data, data_label=None, format=None, show_in_v
+    spectra_to_combine = defaultdict(list)
+
+    for data in datasets:
+        filename = data.meta.get("FILENAME")
```
So with this logic, if there is not `FILENAME` in `data.meta`, then everything will get dumped into the same list. Is that what you want? I guess that is okay but I just want to make sure.

bmorris3 replied:
That's what I was going for, but I would understand if you'd prefer to see an error/warning raised instead.

pllim replied:
No, if this is intentional, it is ok.

pllim reviewed (commented):
I left a few more comments but I don't think I am qualified to approve. I'll delegate that to the devs who have worked on this parser more than I did.

pllim commented:
Do we also want to advertise this new functionality in the user facing doc for Specviz?

bmorris3 commented:
> Do we also want to advertise this new functionality in the user facing doc for Specviz?

I'd love to hear if @orifox has a good example spectrum to use for a concatenation demo in the docs.

pllim commented:
To be clear, I didn't mean a full working demo. I was thinking more like partial snippet to explain that this is an option and when one should use it somewhere on this page:

https://jdaviz.readthedocs.io/en/latest/specviz/import_data.html#importing-data-via-the-api

pllim reviewed (approved):
LGTM. Thanks!

pllim commented:
I wouldn't worry about devdeps but looks like someone just restarted it. LoL.

Feel free to "Squash and Merge" when it passes. Thanks!
